### PR TITLE
Support high power units

### PR DESCRIPTION
**Pull request by Lohrer**
Tracked in #338

### CODE REVIEW

**Comment by linknum23** [comment]
Lmk when it's ready

File: amplipi/rt.py [hunk 1]
```diff
@@ -355,14 +357,26 @@ def _fix2temp(fval: int) -> float:
       temp = fval/2 - 20
     return temp
 
-  def read_temps(self, preamp: int = 1) -> Tuple[Union[float, None], Union[float, None], Union[float, None]]:
+  def hv2_present(self, preamp: int = 1) -> Union[bool, None]:
```
Comment (linknum23) [inline amplipi/rt.py:360]:
```suggestion
  def hv2_present(self, preamp: int = 1) -> Optional[bool]:
```

File: amplipi/rt.py [hunk 2]
```diff
@@ -373,24 +387,35 @@ def read_temps(self, preamp: int = 1) -> Tuple[Union[float, None], Union[float,
...
+      if self.hv2_present(preamp):
+        temp_hv2_f = self.bus.read_byte_data(preamp*8, _REG_ADDRS['HV2_TEMP'])
+        temp_hv2 = self._fix2temp(temp_hv2_f)
+        return temp_hv1, temp_hv2, temp_amp1, temp_amp2
+      else:
+        return temp_hv1, None, temp_amp1, temp_amp2
+    return None, None, None, None
 
-  def read_hv(self, preamp: int = 1) -> Union[float, None]:
+  def read_hv(self, preamp: int = 1) -> Union[Tuple[float, float], Tuple[float, None], Tuple[None, None]]:
```
Comment (linknum23) [inline amplipi/rt.py:398]:
```suggestion
  def read_hv(self, preamp: int = 1) -> Tuple[Optional[float], Optional[float]]:
```
Although this signature isn't as specific it is specific enough and much easer to read

File: fw/preamp/src/adc.c [hunk 3]
```diff
@@ -53,143 +97,192 @@ const uint8_t THERM_LUT_[] = {
...
-
-  while (I2C_GetFlagStatus(I2C2, I2C_FLAG_RXNE) == RESET) {}
-  vals.hv1_temp = I2C_ReceiveData(I2C2);
-
-  while (I2C_GetFlagStatus(I2C2, I2C_FLAG_RXNE) == RESET) {}
-  vals.amp_temp2 = I2C_ReceiveData(I2C2);
-
-  while (I2C_GetFlagStatus(I2C2, I2C_FLAG_STOPF) == RESET) {}
-  I2C_ClearFlag(I2C2, I2C_FLAG_STOPF);
+  for (uint32_t i = 0; i < adc->nchans; i++) {
```
Comment (linknum23) [inline fw/preamp/src/adc.c:167]: I am confused about how this works?

File: fw/preamp/src/adc.c [hunk 4]
```diff
@@ -53,143 +97,192 @@ const uint8_t THERM_LUT_[] = {
...
+    }
+  }
+  return &vals;
 }
 
 void updateAdc() {
 #define ADC_REF_VOLTS 3.3
 #define ADC_PD_KOHMS  4700
 #define ADC_PU_KOHMS  100000
+#define ADC_BITS      8
```
Comment (linknum23) [inline fw/preamp/src/adc.c:244]: Should these be undefed at the bottom of the method?

**Review by linknum23 — APPROVED** [review]
This all looks pretty good. Nice work.

File: fw/preamp/src/adc.h [hunk 5]
```diff
@@ -22,20 +22,54 @@
...
 
+#include <stdbool.h>
 #include <stdint.h>
 
-void initAdc();
-void updateAdc();
+typedef union {
+  // All temps in UQ7.1 + 20 degC format
+  struct {
+    uint8_t hv1_f1;   // PSU 1 temp (always present)
```
[the remark this answers is not in the document]
Comment (Lohrer) [reply fw/preamp/src/adc.h:31]: Yeah I was originally using Q notation but between it the super long names and it being kinda ambiguous for signed types I just stuck with noting the number of fractional bits. Leave it to the IDE to give you data type so you can see the signedness and number of total bits.

**Pull request merged**